### PR TITLE
Add support to shared component

### PR DESCRIPTION
**Pull request by linr211**

### CODE REVIEW

File: src/main/java/com/vmware/vipclient/i18n/I18nFactory.java [hunk 1]
```diff
@@ -98,6 +105,9 @@ public Message getMessageInstance(Class c) {
 		return i;
 	}
 
+	public Message getMessageInstance(Class c) {
+		return this.getMessageInstance(c, null);
```
Comment (Xiaochao8) [inline src/main/java/com/vmware/vipclient/i18n/I18nFactory.java]: Send main config here instead of null.

Comment (linr211) [reply src/main/java/com/vmware/vipclient/i18n/I18nFactory.java]: fixed

File: src/test/java/com/vmware/vip/i18n/SharedComponentTest.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        String comp2 = "JSP";
+        String key2 = "table.head";
+        String source2 = "VM";
+        String trans2 = subTranslation.getString(zhLocale2, comp2, key2, source2, "");
+        logger.debug("pseudoTrans1: " + trans2);
+        Assert.assertTrue(VIPCfg.getInstance().getProductName().equals(mainProductName));
+        Assert.assertTrue(subTranslation.getCfg().getProductName().equals(subProductName));
+
+    }
+}
```
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/SharedComponentTest.java]: Add some tests to test cache.

File: src/test/java/com/vmware/vip/i18n/TranslationMessage2Test.java [hunk 3]
```diff
@@ -27,7 +28,11 @@
 	@Before
 	public void init() {
 		VIPCfg gc = VIPCfg.getInstance();
-		gc.initialize("vipconfig");
+		try {
+			gc.initialize("vipconfig");
+		} catch (VIPClientInitException e) {
+			e.printStackTrace();
```
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/TranslationMessage2Test.java]: Log this message with logger.

File: src/test/java/com/vmware/vip/i18n/VIPCfgTest.java [hunk 4]
```diff
@@ -25,7 +26,11 @@
 	@Before
 	public void init() {
         VIPCfg gc = VIPCfg.getInstance();
-        gc.initialize("vipconfig");
+        try {
+            gc.initialize("vipconfig");
+        } catch (VIPClientInitException e) {
+            e.printStackTrace();
```
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/VIPCfgTest.java]: Log this message with logger.

File: src/test/java/com/vmware/vip/i18n/TranslationMessageTest.java [hunk 5]
```diff
@@ -34,7 +35,11 @@
     @Before
 	public void init() {
         VIPCfg gc = VIPCfg.getInstance();
-        gc.initialize("vipconfig");
+        try {
+            gc.initialize("vipconfig");
+        } catch (VIPClientInitException e) {
+            e.printStackTrace();
```
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/TranslationMessageTest.java]: Log this message with logger.

File: src/test/java/com/vmware/vip/i18n/VIPCfgTest.java (hunk 4 above)
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/VIPCfgTest.java]: Log this message with logger.

File: src/main/java/com/vmware/vipclient/i18n/filters/VIPComponentFilter.java [hunk 6]
```diff
@@ -103,7 +104,11 @@ public void destroy() {
 	private VIPCfg gc = VIPCfg.getInstance();
 	public void init(FilterConfig filterConfig) throws ServletException {
 		if(gc.getVipService() == null) {
-			gc.initialize("vipconfig");
+			try {
+				gc.initialize("vipconfig");
+			} catch (VIPClientInitException e) {
+				e.printStackTrace();
```
Comment (Xiaochao8) [inline src/main/java/com/vmware/vipclient/i18n/filters/VIPComponentFilter.java]: Log this message with logger.

File: src/main/java/com/vmware/vipclient/i18n/fmt/common/MessageSupport.java [hunk 7]
```diff
@@ -46,7 +47,11 @@ private void init() {
 		this.keyAttrValue = null;
 		this.keySpecified = false;
 		VIPCfg gc = VIPCfg.getInstance();
-		gc.initialize("vipconfig");
+		try {
+			gc.initialize("vipconfig");
+		} catch (VIPClientInitException e) {
+			e.printStackTrace();
```
Comment (Xiaochao8) [inline src/main/java/com/vmware/vipclient/i18n/fmt/common/MessageSupport.java]: Log this message with logger.

File: src/test/java/com/vmware/vip/i18n/CacheManagerTest.java [hunk 8]
```diff
@@ -27,7 +28,11 @@
 	@Before
 	public void init() {
 		VIPCfg gc = VIPCfg.getInstance();
-		gc.initialize("vipconfig");
+		try {
+			gc.initialize("vipconfig");
+		} catch (VIPClientInitException e) {
+			e.printStackTrace();
```
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/CacheManagerTest.java]: Log this message with logger.

File: src/test/java/com/vmware/vip/i18n/DateFormatUtilTest.java [hunk 9]
```diff
@@ -22,8 +23,12 @@
 	@Before
 	public void init() {
         VIPCfg gc = VIPCfg.getInstance();
-        gc.initialize("vipconfig");
-        gc.initializeVIPService();
+		try {
+			gc.initialize("vipconfig");
+		} catch (VIPClientInitException e) {
+			e.printStackTrace();
```
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/DateFormatUtilTest.java]: Log this message with logger.

File: src/test/java/com/vmware/vip/i18n/FormatTest.java [hunk 10]
```diff
@@ -18,8 +19,12 @@
 	@Before
 	public void init() {
         VIPCfg gc = VIPCfg.getInstance();
-        gc.initialize("vipconfig");
-        gc.initializeVIPService();
+		try {
+			gc.initialize("vipconfig");
+		} catch (VIPClientInitException e) {
+			e.printStackTrace();
```
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/FormatTest.java]: Log this message with logger.

File: src/test/java/com/vmware/vip/i18n/HttpRequesterTest.java [hunk 11]
```diff
@@ -32,7 +33,11 @@
 	@Before
 	public void init() {
 		VIPCfg cfg = VIPCfg.getInstance();
-		cfg.initialize("vipconfig");
+		try {
+			cfg.initialize("vipconfig");
+		} catch (VIPClientInitException e) {
+			e.printStackTrace();
```
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/HttpRequesterTest.java]: Log this message with logger.

File: src/test/java/com/vmware/vip/i18n/LocaleTest.java [hunk 12]
```diff
@@ -30,8 +31,12 @@
 	@Before
 	public void init() {
         VIPCfg gc = VIPCfg.getInstance();
-        gc.initialize("vipconfig");
-        gc.initializeVIPService();
+		try {
+			gc.initialize("vipconfig");
+		} catch (VIPClientInitException e) {
+			e.printStackTrace();
```
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/LocaleTest.java]: Log this message with logger.

File: src/test/java/com/vmware/vip/i18n/MessageCache2Test1.java [hunk 13]
```diff
@@ -26,7 +27,11 @@
 	@Before
 	public void init() {
 		VIPCfg gc = VIPCfg.getInstance();
-		gc.initialize("vipconfig");
+		try {
+			gc.initialize("vipconfig");
+		} catch (VIPClientInitException e) {
+			e.printStackTrace();
```
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/MessageCache2Test1.java]: Log this message with logger.

File: src/test/java/com/vmware/vip/i18n/MessageCache2Test2.java [hunk 14]
```diff
@@ -26,7 +27,11 @@
 	@Before
 	public void init() {
 		VIPCfg gc = VIPCfg.getInstance();
-		gc.initialize("vipconfig");
+		try {
+			gc.initialize("vipconfig");
+		} catch (VIPClientInitException e) {
+			e.printStackTrace();
```
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/MessageCache2Test2.java]: Log this message with logger.

File: src/test/java/com/vmware/vip/i18n/MessageCacheTest1.java [hunk 15]
```diff
@@ -26,7 +27,11 @@
 	@Before
 	public void init() {
 		VIPCfg gc = VIPCfg.getInstance();
-		gc.initialize("vipconfig");
+		try {
+			gc.initialize("vipconfig");
+		} catch (VIPClientInitException e) {
+			e.printStackTrace();
```
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/MessageCacheTest1.java]: Log this message with logger.

File: src/test/java/com/vmware/vip/i18n/MessageCacheTest2.java [hunk 16]
```diff
@@ -26,7 +27,11 @@
 	@Before
 	public void init() {
 		VIPCfg gc = VIPCfg.getInstance();
-		gc.initialize("vipconfig");
+		try {
+			gc.initialize("vipconfig");
+		} catch (VIPClientInitException e) {
+			e.printStackTrace();
```
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/MessageCacheTest2.java]: Log this message with logger.

File: src/test/java/com/vmware/vip/i18n/NumberFormatUtilTest.java [hunk 17]
```diff
@@ -22,8 +23,12 @@
 	@Before
 	public void init() {
         VIPCfg gc = VIPCfg.getInstance();
-        gc.initialize("vipconfig");
-        gc.initializeVIPService();
+		try {
+			gc.initialize("vipconfig");
+		} catch (VIPClientInitException e) {
+			e.printStackTrace();
```
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/NumberFormatUtilTest.java]: Log this message with logger.

File: src/test/java/com/vmware/vip/i18n/PersistantCacheTest.java [hunk 18]
```diff
@@ -26,7 +27,11 @@
 	@Before
 	public void init() {
 		VIPCfg gc = VIPCfg.getInstance();
-		gc.initialize("vipconfig");
+		try {
+			gc.initialize("vipconfig");
+		} catch (VIPClientInitException e) {
+			e.printStackTrace();
```
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/PersistantCacheTest.java]: Log this message with logger.

File: src/test/java/com/vmware/vip/i18n/PluralFormatTest.java [hunk 19]
```diff
@@ -25,8 +26,12 @@ public PluralFormatTest() {
 	@Before
 	public void init() {
         VIPCfg gc = VIPCfg.getInstance();
-        gc.initialize("vipconfig");
-        gc.initializeVIPService();
+		try {
+			gc.initialize("vipconfig");
+		} catch (VIPClientInitException e) {
+			e.printStackTrace();
```
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/PluralFormatTest.java]: Log this message with logger.

File: src/test/java/com/vmware/vip/i18n/SharedComponentTest.java [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+    String mainProductName = "JavaclientTest";
+    String subProductName = "JavaclientTest1";
+
+    @Before
+    public void init() {
+        VIPCfg mainCfg = VIPCfg.getInstance();
+        try {
+            mainCfg.initialize("vipconfig");
+        } catch (VIPClientInitException e) {
+            e.printStackTrace();
```
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/SharedComponentTest.java]: Log this message with logger.

File: src/test/java/com/vmware/vip/i18n/SharedComponentTest.java [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+        mainCfg.createTranslationCache(MessageCache.class);
+        mainCfg.createFormattingCache(FormattingCache.class);
+        I18nFactory i18n = I18nFactory.getInstance(mainCfg);
+        mainTranslation = (TranslationMessage) i18n.getMessageInstance(TranslationMessage.class);
+
+        VIPCfg subCfg = VIPCfg.getSubInstance(subProductName);
+        try {
+            subCfg.initialize("vipconfig-child");
+        } catch (VIPClientInitException e) {
+            e.printStackTrace();
```
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/SharedComponentTest.java]: Log this message with logger.

File: src/test/java/com/vmware/vip/i18n/MessageFormatTest.java [hunk 22]
```diff
@@ -23,7 +24,11 @@
 	@Before
 	public void init() {
         VIPCfg gc = VIPCfg.getInstance();
-        gc.initialize("vipconfig");
+        try {
+            gc.initialize("vipconfig");
+        } catch (VIPClientInitException e) {
+            e.printStackTrace();
```
Comment (Xiaochao8) [inline src/test/java/com/vmware/vip/i18n/MessageFormatTest.java]: Log this message with logger.

File: src/main/java/com/vmware/vipclient/i18n/I18nFactory.java [hunk 23]
```diff
@@ -32,6 +33,8 @@
 	// store Formatting instance
 	private Map<String, Formatting> formattings = new HashMap<String, Formatting>();
 
+	public static  Map<String, VIPCfg> mapCfg = new HashMap<String, VIPCfg>();
```
Comment (Xiaochao8) [inline src/main/java/com/vmware/vipclient/i18n/I18nFactory.java]: Unused code should be removed.

File: src/test/java/com/vmware/vip/i18n/TranslationMessage2Test.java (hunk 3 above)
Comment (linr211) [reply src/test/java/com/vmware/vip/i18n/TranslationMessage2Test.java]: fixed

File: src/test/java/com/vmware/vip/i18n/TranslationMessageTest.java (hunk 5 above)
Comment (linr211) [reply src/test/java/com/vmware/vip/i18n/TranslationMessageTest.java]: fixed

File: src/test/java/com/vmware/vip/i18n/VIPCfgTest.java (hunk 4 above)
Comment (linr211) [reply src/test/java/com/vmware/vip/i18n/VIPCfgTest.java]: fixed

Comment (linr211) [reply src/test/java/com/vmware/vip/i18n/VIPCfgTest.java]: fixed

File: src/main/java/com/vmware/vipclient/i18n/filters/VIPComponentFilter.java (hunk 6 above)
Comment (linr211) [reply src/main/java/com/vmware/vipclient/i18n/filters/VIPComponentFilter.java]: fixed

File: src/main/java/com/vmware/vipclient/i18n/fmt/common/MessageSupport.java [hunk 24]
```diff
@@ -46,7 +47,11 @@ private void init() {
 		this.keyAttrValue = null;
 		this.keySpecified = false;
 		VIPCfg gc = VIPCfg.getInstance();
-		gc.initialize("vipconfig");
+		try {
+			gc.initialize("vipconfig");
+		} catch (VIPClientInitException e) {
+			
+		}
```
Comment (tigershi) [inline src/main/java/com/vmware/vipclient/i18n/fmt/common/MessageSupport.java]: you can add  exception information log

Comment (linr211) [reply src/main/java/com/vmware/vipclient/i18n/fmt/common/MessageSupport.java]: the fmt doesn't support log.

**Pull request merged**